scc/light_client/light_client_state/light_client_state_test.go

```diff
@@ -0,0 +1,299 @@
+			prov.EXPECT().
+				GetCommitteeCertificates(scc.Period(1), uint64(1)).
+				Return(nil, fmt.Errorf("failed to get committee certificates"))
+		},
+	}
+
+	for name, expectedCalls := range tests {
+		t.Run(name, func(t *testing.T) {
+			prov := provider.NewMockProvider(gomock.NewController(t))
+			state := NewState(scc.Committee{})
+			expectedCalls(prov)
+			_, err := state.Sync(prov)
+			require.ErrorContains(err, "failed to get")
+		})
+	}
+}
+
+func TestLightClientState_Sync_ChangesNothingWhen_LatestBlockIsEmpty(t *testing.T) {
+	require := require.New(t)
+	ctrl := gomock.NewController(t)
+	prov := provider.NewMockProvider(ctrl)
+
+	prov.EXPECT().
+		GetBlockCertificates(provider.LatestBlock, uint64(1)).
+		Return([]cert.BlockCertificate{}, nil)
+
+	state := NewState(scc.Committee{})
+	_, err := state.Sync(prov)
+	require.ErrorContains(err, "zero block certificates")
+	want := State{}
+	compareStates(t, &want, state)
+}
+
+func TestLightClientState_Sync_UpdatesOnlyHeadWhen_SyncToCurrentPeriod(t *testing.T) {
+	require := require.New(t)
+	ctrl := gomock.NewController(t)
+	prov := provider.NewMockProvider(ctrl)
+
+	// setup state with non empty committee
+	key := bls.NewPrivateKey()
+	member := makeMember(key)
+	state := NewState(scc.NewCommittee(member))
+	state.period = 1
+
+	// setup block for period 1.
+	blockNumber := idx.Block(scc.BLOCKS_PER_PERIOD*1 + 1)
+	blockCert := cert.NewCertificate(
+		cert.NewBlockStatement(0, blockNumber, common.Hash{0x1}, common.Hash{0x2}))
+
+	// sigh the block certificate with the committee member
+	err := blockCert.Add(scc.MemberId(0), cert.Sign(blockCert.Subject(), key))
+	require.NoError(err)
+
+	prov.EXPECT().
+		GetBlockCertificates(provider.LatestBlock, uint64(1)).
+		Return([]cert.BlockCertificate{blockCert}, nil)
+
+	_, err = state.Sync(prov)
+	require.NoError(err)
+	want := State{
+		period:     scc.Period(1),
+		committee:  scc.NewCommittee(member),
+		headNumber: blockCert.Subject().Number,
+		headHash:   common.Hash{0x1},
+		headRoot:   common.Hash{0x2},
+	}
+	compareStates(t, &want, state)
+}
+
+func TestLightClientState_Sync_CanNotSyncToPastPeriod(t *testing.T) {
+	require := require.New(t)
+	ctrl := gomock.NewController(t)
+	prov := provider.NewMockProvider(ctrl)
+
+	// setup block for period 1.
+	expectBlockForPeriod(prov, 1)
+
+	// setup synced to period 2
+	state := State{period: 2}
+	_, err := state.Sync(prov)
+	require.ErrorContains(err, "cannot sync to a previous period")
+}
+
+func TestLightClientState_Sync_ReportsFailedVerificationOfLatestBlock(t *testing.T) {
+	require := require.New(t)
+	ctrl := gomock.NewController(t)
+	prov := provider.NewMockProvider(ctrl)
+
+	// setup state with non empty committee
+	key := bls.NewPrivateKey()
+	member := makeMember(key)
+	state := NewState(scc.NewCommittee(member))
+	state.period = 1
+
+	// setup unsigned block for period 1.
+	expectBlockForPeriod(prov, 1)
+
+	_, err := state.Sync(prov)
+	require.ErrorContains(err, "insufficient voting power")
```

Comment:
Two style remarks here:
- this error message is produced by the `cert` package and should not be tested here; changes of the `cert` package to under-specified features should not cause tests in other packages to fail
- it might be good to provide an error message that fits the context of the state; something like "failed to authenticate block certificate for head block" ...

Reply:
https://github.com/0xsoniclabs/sonic/pull/100/commits/2a770df297b2db0fd37431b481c0f6b7c81a9682